server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -535,50 +615,238 @@ protected Releasable drainSync() {
+                    }
+
+                    // Check last fetch status of pinned timestamps. If stale, return.
+                    if (indexDeleted == false && RemoteStoreUtils.isPinnedTimestampStateStale()) {
+                        logger.warn("Skipping remote segment store garbage collection as last fetch of pinned timestamp is stale");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    List<String> metadataFilesToBeDeleted = getMetadataFilesToBeDeleted(metadataFiles);
+
+                    // If index is not deleted, make sure to keep latest metadata file
+                    if (indexDeleted == false) {
+                        metadataFilesToBeDeleted.remove(metadataFiles.get(0));
+                    }
+
+                    if (metadataFilesToBeDeleted.isEmpty()) {
+                        logger.debug("No metadata files to delete");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    // For all the files that we are keeping, fetch min and max generations
+                    List<String> metadataFilesNotToBeDeleted = new ArrayList<>(metadataFiles);
+                    metadataFilesNotToBeDeleted.removeAll(metadataFilesToBeDeleted);
+
+                    Set<Long> generationsToBeDeleted = getGenerationsToBeDeleted(
+                        metadataFilesNotToBeDeleted,
+                        metadataFilesToBeDeleted,
+                        indexDeleted
+                    );
+                    if (generationsToBeDeleted.isEmpty() == false) {
+                        // Delete stale generations
+                        translogTransferManager.deleteGenerationAsync(
+                            primaryTermSupplier.getAsLong(),
+                            generationsToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Delete stale metadata files
+                        translogTransferManager.deleteMetadataFilesAsync(
+                            metadataFilesToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Update cache to keep only those metadata files that are not getting deleted
+                        oldFormatMetadataFileGenerationMap.keySet().retainAll(metadataFilesNotToBeDeleted);
+
+                        // Delete stale primary terms
+                        deleteStaleRemotePrimaryTerms(metadataFiles);
+                    } else {
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                    }
+                } catch (Exception e) {
+                    remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                }
```

Comment:
Use a finally block instead?

Reply:
Actually, there are 2 permits and translog and metadata generation releases 1 each. This way, we make sure to release permit only after deletion is completed.

Reply:
But currently, metadata and generation delete handles the permit release concurrently. I think we can make it sequential.